fog/overseer/server/src/worker.rs

```diff
@@ -0,0 +1,511 @@
+                    );
+                    ingest_summary_node_mappings.push(IngestSummaryNodeMapping {
+                        node_index: ingest_client_index,
+                        ingest_summary,
+                    });
+                }
+                Err(_) => {
+                    let error_message = format!(
+                        "Unable to retrieve ingest summary for node: {}",
+                        ingest_client.get_uri()
+                    );
+                    return Err(OverseerError::UnresponsiveNodeError(error_message));
+                }
+            }
+        }
+
+        Ok(ingest_summary_node_mappings)
+    }
+
+    /// Performs automatic failover, which means that we try to activate nodes
+    /// for an outstanding ingress key, if it exists.
+    ///
+    /// The logic is as follows:
+    ///   1. Find all of the "outstanding keys" as determined by the RecoveryDb.
+    ///      These are ingress keys that Fog Ingest needs to scan blocks with
+    ///      but Fog Ingest isn't currently doing that because all nodes are
+    ///      idle.
+    ///   2. If there are:
+    ///         a) 0 outstanding keys:
+    ///              No node will be activated. We now have to:
+    ///                 (i)  Set new keys on an idle node.
+    ///                 (ii) Activate that node.
+    ///         b) 1 oustanding key:
+    ///              Try to find an idle node that contains that key.
+    ///                 (i)  If you find one, great! Just activate that node. If
+    ///                      activation is unsuccessful, then return an error
+    ///                      and return to the overseer polling logic.
+    ///                 (ii) If you don't find an idle node with that key,
+    ///                      then you have to report that key as lost, set
+    ///                      new keys on an idle node, and activate that node.
+    ///        c) > 1 outstanding key:
+    ///             (i) Disable
+    ///             (ii) TODO: Send an alert.
+    fn perform_automatic_failover(
+        &self,
+        ingest_summary_node_mappings: Vec<IngestSummaryNodeMapping>,
+    ) -> Result<(), OverseerError> {
+        let inactive_outstanding_keys: Vec<CompressedRistrettoPublic> =
+            self.get_inactive_outstanding_keys()?;
+
+        match inactive_outstanding_keys.len() {
+            0 => log::info!(self.logger, "Found 0 outstanding keys."),
+            1 => log::info!(self.logger, "Found 1 outstanding key."),
+            _ => {
+                log::error!(self.logger, "Found multiple outstanding keys. This requires manual intervention. Disabling overseer.");
+                self.is_enabled.store(false, Ordering::SeqCst);
+                return Err(OverseerError::MultipleOutstandingKeys("Multiple outstanding keys found. This is unexpected and requires manual intervention. As such, we've disabled overseer. Take action and then enable overseer.".to_string()));
```

Comment:
Nit: Maybe be worth including the keys?

Reply:
Another nit is I've seen a mixture of `.to_string()`, `.to_owned()` and `String::from()`. They all do the same, I think `to_owned()` or `to_string()` are both used in the codebase, I would pick one and stick to it.

Reply:
I included the keys and standardized on to_string.